### ISSUE TITLE
Context Selectors vs Proxy Handlers

### ISSUE THREAD

**Issue by joepuzzo** [dai-shi/react-tracked#94]
So the concept of reactSelectors was recently introduced as a solution too obvious excessive rendering issues. 

```js
<SomeContext.Provider value={{
  foo: 'Hello',
  bar: 'World'
}}>
 ...
</SomeContext.Provider>

// Before Context selectors
const { foo } = useContext(SomeContext)

// With Context Selectors
const foo = useContextSelector(SomeContext, v => v.foo )
```

The solution above is highly EXPLICIT about what will cause a re render.

However, now we have this solution, react-tracked, which fundamentally takes a different approach.

```js
<SomeContext.Provider value={{
  foo: 'Hello',
  bar: 'World'
}}>
 ...
</SomeContext.Provider>

// Before react-tracked
const { foo } = useContext(SomeContext)

// With  react-tracked ( pseudo code )
const { foo } = useMagicContext(SomeContext)
```

Problem with this ^^^^^^ is its not explicit, and can lead to confusion.

For example: 

```js
const context = useMagicContext(SomeContext)

// This will only render when foo changes *Awesome!
return (
  <span>{context.foo}</span>
}
```

BUT, then the user adds one line of code

```js
const context = useMagicContext(SomeContext)

console.log(context);

// This will render when anything on context changes
return (
  <span>{context.foo}</span>
}
```

In other words what the user does on the LEFT_HAND side of the operand will change how the component renders. 

VS 

With context selectors, the right hand side controls what triggers a render.

MY QUESTION: do you see issues with this "tracked" approach being too magical? I.E not explicit enough?

**Comment by dai-shi** [comment]
I think it's educational/documentation issues. It is magical without knowing it's tracked. That's why I explicitly named the hook and the library "tracked".
Do you have other examples that fall into this problem than console.log? There must be, but I'd like to learn.

I know some people prefer to avoid magical tracking with proxies. That's fair. For those people, I'd recommend [jotai](https://github.com/pmndrs/jotai). All my efforts around these libraries are to avoid selectors for optimization (selectors for semantics are fine). jotai is one without proxies. [proxy-memoize](https://github.com/dai-shi/proxy-memoize) is another library for the same goal with proxies, which you might prefer as it on the RIGHT_HAND side.

**Comment by joepuzzo** [comment]
Yeah how it happens under the hood I guess I could care less about. I have written my own implementation of context selectors as well prior to seeing react-tracked. After seeing this however, It sent me down a rabbit hole. Contemplating what is the better approach. 

Im really curious to see how the react core team solves this problem. No matter what, we need some solution the excessive rendering problem. 

IMO, its better to be explicit about when the component will re-render. Technically, if there was a "mind-shift" on how re-renders get triggered, then reading the code: 

```js
return (
  <span>{context.foo}</span>
}
```

would make sense because I could read that code as `"Rerender becausea you rendered context.foo and it changed"`

as apposed to how we read react code today `"Rerender becausea you rendered context.foo and context changed"`

```js
const { foo } = useContext(MyContext)

return (
  <span>{context.foo}</span>
}
```

**Comment by joepuzzo** [comment]
I will see if I can come up with better examples. But my concern with "magic" code is that people may scratch there heads trying to figure out why simply adding a console.log statement changed how often their component re-renderd lol

**Comment by dai-shi** [comment]
BTW, I agree the console.log is a real pitfall.
It would be nice to document the recommended usage for console.log.
As we should avoid console.log in render in pure react, it would be like this:

```js
// if you want to console.log a tracked object, do like this.

import { getUntrackedObject } from 'react-tracked';

const Component = () => {
  const state = useTrackedState();
  useEffect(() => {
    console.log(getUntrackedObject(state));
  })
  // ...
};
```

**Comment by joepuzzo** [comment]
Right. The key here is that users might forget that they have optimized proxy get under the hood and then get confused. Another example would be this. I maintain a react form library, and sometimes a user will write this.

```jsx

const MyComponent = () => {
   const { values } = useFormState();

   return (
      <React.Fragment>
            <span>{values.foo}</span>
            <span>{values.bar}</span>
      </React.Fragment>
   );
}
```

A simple enough component. Lets now suppose that under the hood it uses a proxy handler. 

But, the user then goes to debug there form and does this. 

```jsx

const MyComponent = () => {
   const { values } = useFormState();

   return (
      <React.Fragment>
            <span>{values.foo}</span>
            <span>{values.bar}</span>
            <pre>{JSON.stringify(values, null, 2)}</pre>
      </React.Fragment>
   );
}
```

All of a sudden, while they are debugging some issue, they start to get more renders than the code without that debug output. The question becomes, is this obvious to the user that it renders more depending on what is "used"

**Comment by joepuzzo** [comment]
All they did is render all the values, instead of just two, without an explicit selector your typical react developer would think they would render whenever values change either way.

**Comment by dai-shi** [comment]
Thanks for another example. Yeah, JSON.stringify would access all properties. I wonder how Vue solves this issue.

This is "proxy leaking" behavior, which some people like, while some people do not. As you said, it requires "mind-shift". For typical react developers who are already familiar with non-magical behavior, proxy-memoize would fit better as it doesn't leak proxies.

**Comment by joepuzzo** [comment]
Technically you could implement proxy handler under the hood, and expose it to the user as selectors. That way its clear. The question is, are proxy handlers the best ( most efficient ) solution to this problem 🤔

**Comment by joepuzzo** [comment]
In the end I feel like the react team will need to do something soon to support this behavior. I see more developers use context every day, ( which is great! ) , but then they don't have appreciation for how react renders, and end up with so much extra re-renders due to throwing more things onto context. The goal is to make it easy for devs to not misuse the tools.

**Comment by axeljeremy7** [comment]
have you test your samples in code https://codesandbox.io/s/reverent-tree-geptx?file=/package.json , i use 1.4.2 because after that there is always re rendering, is anyone aware of this? or I have to switch to jotai

**Comment by dai-shi** [comment]
@axeljeremy7 Do you mean "extra render function invocation without commit" behavior? If that's the case, yes, it's introduced in v1.5.0 which depends on use-context-selector with the behavior, and later made it opt-in in v1.7.0. Would you please try the latest version? ref: #90, #92.

**Comment by axeljeremy7** [comment]
latest version with the codesandbox has same behavior or I have to use other library or add library.

**Comment by dai-shi** [comment]
I'm not sure if I understand how to reproduce a different behavior.
https://codesandbox.io/s/gifted-cdn-x9fqd
Switching v1.7.3 and v1.4.2 doesn't seem to defer as far as `useFlasher` tells.
Would you please explain how to reproduce?

**Comment by axeljeremy7** [comment]
I have an issue with the lib, if update it increments 50 commits based on the profiler and the other one I found out, is that the arrays in my state are re rendering more than once.

```
const state = useTrackedState();
    // @ts-ignore
    const {client, pinChannels, groupChannels, duoChannels, loading} = state;
```
https://www.dropbox.com/s/h8egyu9vckp6zxe/Screen%20Shot%202021-07-28%20at%2012.50.32%20PM.png?dl=0

somehow the arrays are causing to renders.

**Comment by dai-shi** [comment]
That sounds like a bug. It would be so appreciated if you could build a minimal reproduction.

**Comment by axeljeremy7** [comment]
no but the problem is that when I dispatch and update part of the state , it renders again but the payload has nothing to do with the arrays , the payload to update doesn't modify the arrays so why is thinking the there has been changes.?

**Comment by dai-shi** [comment]
If it's hard to provide a reproduction, you can try debugging by yourself with `DebugValue`.
Look into `useTrackedState` and find `pathList` which shows tracking info (= what triggers re-renders.)

**Comment by axeljeremy7** [comment]
https://github.com/axeljeremy7/stateManagementReact/tree/main/react_tracked_test
https://codesandbox.io/s/sweet-johnson-e0upi?file=/package.json

I create a repo to reproduce my issues.
The problem is that when I click on generate new random number(see the repo to understand), the TodoList is re rendered , and in my app, each row have a lot components so I can't afford. I want to avoid the re render on the list.

**Comment by dai-shi** [comment]
If I understand correctly, you are referring to `'rendered TodoList'` log, when clicking `Generate My Random Number` button.
This is because you use the random number and show it in TodoList.

https://codesandbox.io/s/immutable-waterfall-vojhq?file=/src/TodoList.tsx
Here's a fix.

If this is not what you are referring, please elaborate the steps to check the behavior.

**Comment by axeljeremy7** [comment]
check the repo please,  I fixed the issue,  the console log doesn't say render anymore, but the profiler it showing "Did not render" when I click even though it shouldn't, I. think it should be just 1 commit instead of 2 commit.

**Comment by dai-shi** [comment]
https://githubbox.com/axeljeremy7/stateManagementReact/tree/main/react_tracked_test
Does this reproduce your issue?

> the profiler it showing "Did not render" 

I'm not very familiar with the profiler. Would you explain how I can see the issue?

Is this the one?
<img width="407" alt="image" src="https://user-images.githubusercontent.com/490574/127720716-b750d15a-00cd-4695-ac50-83de81029265.png">

> it should be just 1 commit instead of 2 commit.

Does 2 commit mean `useEffect(() => { console.log('commit'); });` shows double logs?

**Comment by dai-shi** [comment]
Closing as stale.
